### ISSUE TITLE
As a network starter, I can specify a HLF v1.0 endorsement policy during network start

### ISSUE THREAD

**Issue by sstone1** [hyperledger/composer#1251]
Hyperledger Fabric v1.0 provides support for endorsement policies:
https://hyperledger-fabric.readthedocs.io/en/latest/endorsement-policies.html

When using the CLI, endorsement policies are a basic DSL that allows a network starter to specify the organizations which must endorse a transaction in order for it to be committed to the ledger.

Example CLI endorsement policy: `AND('Org1.member', 'Org2.member')`

This basic DSL is not available in the Node.js SDK nor the Java SDK. Instead, there is a JSON/YAML serialization format which allows for more expressive endorsement policies. For example:

```
{
   "identities": [
      {
         "role": {
            "name": "member",
            "mspId": "peerOrg1"
         }
      },
      {
         "role": {
            "name": "member",
            "mspId": "peerOrg2"
         }
      },
      {
         "role": {
            "name": "admin",
            "mspId": "ordererOrg"
         }
      }
   ],
   "policy": {
      "2-of": [
         {
            "signed-by": 2
         },
         {
            "1-of": [
               {
                  "signed-by": 0
               },
               {
                  "signed-by": 1
               }
            ]
         }
      ]
   }
}
```

Endorsement policies are an operational concern; they do not influence the programming model, and must be built by somebody who understands the structure of the underlying Hyperledger Fabric network. Additionally, the same business network definition can be deployed with different endorsement policies depending on the environment that they are deployed into. For that reason, they do not belong in the business network definition/archive.

Endorsement policies are also specific to Hyperledger Fabric v1.0, and because Composer is designed to be capable of working with a range of Blockchain platforms/versions, we do not want to expose these directly via the Composer APIs or CLIs.

Instead, we need to introduce the ability to pass connector specific options through to the deploy commands via the APIs and the CLIs. I propose that we do this as follows:

`Connection.deploy(securityContext, force, businessNetwork)` is extended to include an additional, optional argument `options` which accepts a JavaScript object containing key/value pairs:
`Connection.deploy(securityContext, force, businessNetwork[, options])`
https://github.com/hyperledger/composer/blob/master/packages/composer-connector-hlfv1/lib/hlfconnection.js#L486 

`AdminConnection.deploy(businessNetworkDefinition)` is extended in a similar way, and just passes the options straight through to the connector:
`AdminConnection.deploy(businessNetworkDefinition[, options])`
https://github.com/hyperledger/composer/blob/master/packages/composer-admin/lib/adminconnection.js#L226

The CLI is updated with two optional arguments that allow these options to be specified via the command line or via a configuration file:
`composer network deploy -p hlfv1 -a biznet.bna -o connectorOption1=someValue -o connectorOption2=someOtherValue`
`composer network deploy -p hlfv1 -a biznet.bna -O connectorOptionFile.txt`

In the case of the first example, the options argument to `AdminConnection.deploy` would look as follows:

```
{
  "connectorOption1":"someValue",
  "connectorOption2":"someOtherValue"
}
```

The HLFv1 connector can then be updated to look for `endorsementPolicy={ ... JSON syntax above ...}` or `endorsementPolicyFile=/path/to/endorsement/policy.json` options and include them in the instantiate request.

**Comment by davidkel** [comment]
@sstone1 should we look to remove the specific force parameter in the Connection interface into the deployOptions object ?

**Comment by davidkel** [comment]
@sstone1 should this also be part of #1390 ?

**Comment by jeffgarratt** [comment]
I would believe this to a very important aspect to the BN, as you can enforce that the network can not proceed without all organizations peers being verified available.  By this, I mean an organization that was for whatever reason experiencing an outage could be assured that progress is not being made.  This would not be important for every scenario, but I would think for many it would be a source of assurance.

**Comment by davidkel** [comment]
```
		// use this to demonstrate the following policy:
		// 'if signed by org1 admin, then that's the only signature required,
		// but if that signature is missing, then the policy can also be fulfilled
		// when members (non-admin) from both orgs signed'
		'endorsement-policy': {
			identities: [
				{ role: { name: 'member', mspId: 'Org1MSP' }},
				{ role: { name: 'member', mspId: 'Org2MSP' }},
				{ role: { name: 'admin', mspId: 'Org1MSP' }}
			],
			policy: {
				'1-of': [
					{ 'signed-by': 2},
					{ '2-of': [{ 'signed-by': 0}, { 'signed-by': 1 }]}
				]
			}
		}
```

**Comment by davidkel** [comment]
Currently the connector checks all responses are valid and that all responses match. Now with endorsement policies the client (without a lot of extra code and maybe not even possible) to know which peer responses are appropriate and which ones need to match (ie if one of the peers RW set isn't the same is that a problem if the others match the endorsement policy). 
So only thing we can do is check each individual responses and throw away invalid ones and send the rest to the orderer for the peers to validate against the policy.
ie remove the RW set check completely

**Comment by davidkel** [comment]
ok, so in summary this is how it will work...
1. composer network deploy -o endorsementPolicy='{"endorsementPolicy": {"identities": [.... }' (JSON in a single line for the endosementPolicy option)
2. composer network deploy -o endorsementPolicyFile=/path/to/file/endorsementPolicy.json
3. composer network deploy -O /path/to/file/options.json

**Comment by davidkel** [comment]
In order to support the -O option which allows for something like this (so we can have a multiline layout
```
{
    "endorsementPolicy": {
        "identities": [
            {
                "role": {
                    "name": "member",
                    "mspId": "Org1MSP"
                }
            },
            {
                "role": {
                    "name": "member",
                    "mspId": "Org2MSP"
                }
            },
            {
                "role": {
                    "name": "admin",
                    "mspId": "Org1MSP"
                }
            }
        ],
        "policy": {
            "1-of": [
                {
                    "signed-by": 2
                },
                {
                    "2-of": [
                        {
                            "signed-by": 0
                        },
                        {
                            "signed-by": 1
                        }
                    ]
                }
            ]
        }
    }
}
```
would be to define the concept of connector specific parameters as allowing either string or object to be passed through and the connector will need to expect either.

**Comment by davidkel** [comment]
@EdProsser adding you to this issue as their will be a documentation impact, so we should get together and discuss.

**Comment by davidkel** [comment]
```
composer network deploy [options]

Options:
  --help                       Show help  [boolean]
  -v, --version                Show version number  [boolean]
  --archiveFile, -a            The business network archive file name  [string] [required]
  --connectionProfileName, -p  The connection profile name  [string]
  --enrollId, -i               The enrollment ID of the user  [string] [required]
  --loglevel, -l               The initial loglevel to set (INFO|WARNING|ERROR|DEBUG)  [string]
  --option, -o                 Options that are specific specific to connection. Multiple options are specified by repeating this option  [string]
  --optionsFile, -O            A file containing options that are specific to connection  [string]
  --enrollSecret, -s           The enrollment secret of the user  [string]
```

**Comment by davidkel** [comment]
now available in 0.9.1
